lib/srv/alpnproxy/proxy.go

```diff
@@ -0,0 +1,352 @@
+}
+
+func (p *Proxy) handleDatabaseConnection(ctx context.Context, conn net.Conn) error {
+	if p.cfg.Router.databaseTLSHandler == nil {
+		return trace.BadParameter("database handle not enabled")
+	}
+	return p.cfg.Router.databaseTLSHandler.Handler(ctx, conn)
+}
+
+func (p *Proxy) databaseHandlerWithTLSTermination(ctx context.Context, conn net.Conn) error {
+	// DB Protocols like Mongo have native support for TLS thus TLS connection needs to be terminated twice.
+	// First time for custom local proxy connection and the second time from Mongo protocol where TLS connection is used.
+
+	tlsConn := tls.Server(conn, p.cfg.TLSConfig)
+	if err := tlsConn.SetReadDeadline(p.cfg.Clock.Now().Add(p.cfg.ReadDeadline)); err != nil {
+		tlsConn.Close()
+		return trace.Wrap(err)
+	}
+	if err := tlsConn.Handshake(); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := tlsConn.SetReadDeadline(time.Time{}); err != nil {
+		tlsConn.Close()
+		return trace.Wrap(err)
+	}
+
+	if !p.isDatabaseConnection(tlsConn.ConnectionState()) {
+		return trace.BadParameter("not database connection")
+	}
+	return trace.Wrap(p.handleDatabaseConnection(ctx, tlsConn))
+}
+
+func isDBTLSProtocol(protocol string) bool {
+	switch protocol {
+	case ProtocolMongoDB:
+		return true
+	default:
+		return false
+	}
+}
+
+func (p *Proxy) getHandlerDescBaseOnClientHelloMsg(clientHelloInfo *tls.ClientHelloInfo) (*HandlerDecs, error) {
+	if shouldRouteToKubeService(clientHelloInfo.ServerName) {
+		if p.cfg.Router.kubeHandler == nil {
+			return nil, trace.BadParameter("received kube request but k8 service is disabled")
+		}
+		return p.cfg.Router.kubeHandler, nil
+	}
+	return p.getHandleDescBasedOnALPNVal(clientHelloInfo)
+}
+
+// getHandleDescBasedOnALPNVal returns the HandlerDesc base on ALPN field read from ClientHelloInfo message.
+func (p *Proxy) getHandleDescBasedOnALPNVal(clientHelloInfo *tls.ClientHelloInfo) (*HandlerDecs, error) {
+	protocol := ProtocolDefault
+	if len(clientHelloInfo.SupportedProtos) != 0 {
+		protocol = clientHelloInfo.SupportedProtos[0]
+	}
+
+	if isDBTLSProtocol(protocol) {
+		return &HandlerDecs{
+			Protocols:  []string{protocol},
+			Handler:    p.databaseHandlerWithTLSTermination,
+			ForwardTLS: false,
+		}, nil
+	}
+
+	handlerDesc, ok := p.cfg.Router.alpnHandlers[protocol]
+	if !ok {
+		return nil, trace.BadParameter("unsupported ALPN protocol %q", protocol)
+	}
+	return handlerDesc, nil
+}
+
+func shouldRouteToKubeService(sni string) bool {
+	return strings.HasPrefix(sni, KubeSNIPrefix)
+}
+
+// Close the Proxy server.
+func (p *Proxy) Close() error {
+	if p.cancel != nil {
+		p.cancel()
+	}
+	if err := p.cfg.Listener.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// isDatabaseConnection inspects the TLS connection state and returns true
+// if it's a database access connection as determined by the decoded
+// identity from the client certificate.
+func (p *Proxy) isDatabaseConnection(state tls.ConnectionState) bool {
```

Comment:
There's the same function in lib/multiplexer/web.go. Can it be factored out and reused in both places? (Unless you removed it from there)

Reply:
Refactored to the utils package